np.ndarray.newbyteorder removed in numpy 2.0

To assist reproducing bugs, please include the following:
 * Operating System (Ubuntu 20.04.3 LTS)
 * Python version (3.10.12)
 * Where Python was acquired (conda-forge)
 * h5py version (3.10.0)
 * HDF5 version (1.14.2)
 * The full traceback/stack trace shown (if it appears)

<details>
    <summary>Trace Back</summary>
____________________ TestH5NetCDFData.test_roundtrip_endian ____________________

self = <xarray.tests.test_backends.TestH5NetCDFData object at 0x7f5542b6c8e0>

    def test_roundtrip_endian(self) -> None:
        ds = Dataset(
            {
                "x": np.arange(3, 10, dtype=">i2"),
                "y": np.arange(3, 20, dtype="<i4"),
                "z": np.arange(3, 30, dtype="=i8"),
                "w": ("x", np.arange(3, 10, dtype=float)),
            }
        )
    
>       with self.roundtrip(ds) as actual:

/home/runner/work/xarray/xarray/xarray/tests/test_backends.py:1087: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/home/runner/micromamba/envs/xarray-tests/lib/python3.10/contextlib.py:135: in __enter__
    return next(self.gen)
/home/runner/work/xarray/xarray/xarray/tests/test_backends.py:309: in roundtrip
    with self.open(path, **open_kwargs) as ds:
/home/runner/micromamba/envs/xarray-tests/lib/python3.10/contextlib.py:135: in __enter__
    return next(self.gen)
/home/runner/work/xarray/xarray/xarray/tests/test_backends.py:335: in open
    with open_dataset(path, engine=self.engine, **kwargs) as ds:
/home/runner/work/xarray/xarray/xarray/backends/api.py:573: in open_dataset
    backend_ds = backend.open_dataset(
/home/runner/work/xarray/xarray/xarray/backends/h5netcdf_.py:414: in open_dataset
    ds = store_entrypoint.open_dataset(
/home/runner/work/xarray/xarray/xarray/backends/store.py:58: in open_dataset
    ds = Dataset(vars, attrs=attrs)
/home/runner/work/xarray/xarray/xarray/core/dataset.py:696: in __init__
    variables, coord_names, dims, indexes, _ = merge_data_and_coords(
/home/runner/work/xarray/xarray/xarray/core/dataset.py:425: in merge_data_and_coords
    return merge_core(
/home/runner/work/xarray/xarray/xarray/core/merge.py:718: in merge_core
    collected = collect_variables_and_indexes(aligned, indexes=indexes)
/home/runner/work/xarray/xarray/xarray/core/merge.py:358: in collect_variables_and_indexes
    variable = as_variable(variable, name=name)
/home/runner/work/xarray/xarray/xarray/core/variable.py:158: in as_variable
    obj = obj.to_index_variable()
/home/runner/work/xarray/xarray/xarray/core/variable.py:552: in to_index_variable
    return IndexVariable(
/home/runner/work/xarray/xarray/xarray/core/variable.py:2632: in __init__
    self._data = PandasIndexingAdapter(self._data)
/home/runner/work/xarray/xarray/xarray/core/indexing.py:1481: in __init__
    self.array = safe_cast_to_index(array)
/home/runner/work/xarray/xarray/xarray/core/indexes.py:469: in safe_cast_to_index
    index = pd.Index(np.asarray(array), **kwargs)
/home/runner/work/xarray/xarray/xarray/core/indexing.py:474: in __array__
    return np.asarray(self.get_duck_array(), dtype=dtype)
/home/runner/work/xarray/xarray/xarray/core/indexing.py:554: in get_duck_array
    array = self.array[self.key]
/home/runner/work/xarray/xarray/xarray/coding/variables.py:112: in __getitem__
    return np.asarray(self.array[key], dtype=self.dtype)
/home/runner/work/xarray/xarray/xarray/core/indexing.py:474: in __array__
    return np.asarray(self.get_duck_array(), dtype=dtype)
/home/runner/work/xarray/xarray/xarray/core/indexing.py:554: in get_duck_array
    array = self.array[self.key]
/home/runner/work/xarray/xarray/xarray/backends/h5netcdf_.py:49: in __getitem__
    return indexing.explicit_indexing_adapter(
/home/runner/work/xarray/xarray/xarray/core/indexing.py:861: in explicit_indexing_adapter
    result = raw_indexing_method(raw_key.tuple)
/home/runner/work/xarray/xarray/xarray/backends/h5netcdf_.py:56: in _getitem
    return array[key]
/home/runner/micromamba/envs/xarray-tests/lib/python3.10/site-packages/h5netcdf/core.py:347: in __getitem__
    return self._h5ds[key]
h5py/_objects.pyx:54: in h5py._objects.with_phil.wrapper
    ???
h5py/_objects.pyx:55: in h5py._objects.with_phil.wrapper
    ???
/home/runner/micromamba/envs/xarray-tests/lib/python3.10/site-packages/h5py/_hl/dataset.py:758: in __getitem__
    return self._fast_reader.read(args)
h5py/_selector.pyx:368: in h5py._selector.Reader.read
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   AttributeError: `newbyteorder` was removed from the ndarray class in NumPy 2.0. Use `arr.view(arr.dtype.newbyteorder(order))` instead.

h5py/_selector.pyx:344: AttributeError
</details>

This was raised within the xarray upstream dev test suite. The issue source is probably this line:

https://github.com/h5py/h5py/blob/f88fb8c5dc5a143037bd5251a7061f6731306209/h5py/_selector.pyx#L344

--- a/h5py/_selector.pyx
+++ b/h5py/_selector.pyx
@@ -341,7 +341,7 @@ cdef class Reader:
 
             arr = PyArray_ZEROS(arr_rank, arr_shape, self.np_typenum, 0)
             if not self.native_byteorder:
-                arr = arr.newbyteorder()
+                arr = arr.view(arr.dtype.newbyteorder())
         finally:
             efree(arr_shape)
 

